### DISCUSSION TITLE
typescript declare, module, namespace

### DISCUSSION THREAD

**Discussion by al-bur** [woowacourse-study/2022-woowahan-bansanghwe#124] [TypeScript]
처음 typescript를 공부하고 프로젝트에 타입스크립트를 적용하다보면 항상 만나는 키워드들이 있다.

`declare`, `module`, `namespace` 

오늘은 해당 키워드들에 대해서 알아보려 한다.

## declare가 뭔데?

---

첫 번째 키워드는 `declare`다.

무엇인가에 대한 정의를 찾아볼 때는 공식문서에서 찾아보는 게 제일 정확하다. 근데, 아쉽게도 정확히 `declare`라는 단어가 어떤 의미를 가지고 있는 지는 알려주고 있지 않아 다른 사이트들을 참고 해보았다.

> The declare keyword in TypeScript is used for the Ambient declaration of variables or for methods. Ambient Declarations is like an import keyword. Which tells the compiler that the source exists in another file.
> 

해석을 해보면, `declare` 키워드는 `ambient declaration` 을 위해 사용된다고 한다. `ambient declaration` 은 구현부가 없는 타입 선언이라고 이해하면 된다. 

## declare namespace vs declare module

---

그럼 이제, namespace와 module의 차이를 알아보자

찾아본 결과, 두 개 모두 모듈을 구성하기 위한 키워드란 것을 알 수 있었다.

**Comment by airman5573** [comment]
딱 간지러운 부분들이네요!! 기대하겠습니다 👍

**Reply by byhhh2** [reply]
간질간질!

**Comment by airman5573** [comment]
가장 최근에 해결했던 문제 해결

**Reply by byhhh2** [reply]
가장 해결하기 어려웠던 문제?

**Comment by airman5573** [comment]
가장 어려웠던 문제 해결

**Comment by byhhh2** [comment]
typescript를 사용한 이유

**Reply by byhhh2** [reply]
typescript를 사용할 때 어려웠던 점

**Comment by byhhh2** [comment]
반응형을 구현하면서 가장 어려웠던 부분?

**Reply by byhhh2** [reply]
`rem`을 사용한 이유 ?

**Reply by byhhh2** [reply]
`rem`을 사용해서 반응형을 해결하지 못한 이유

**Reply by byhhh2** [reply]
원인 분석은 어떻게 하셨나요?

**Comment by airman5573** [comment]
현재 팀 컨벤션이 어떻게 진행 될까요?

**Comment by airman5573** [comment]
왜 REM을 사용하고 -> 반응형 -> 왜 튀어나왔는지?

**Comment by airman5573** [comment]
테스트 도입률

**Comment by airman5573** [comment]
자바스크립트를 공부하면서 가장 어려웠던 부분은?

**Comment by airman5573** [comment]
실행컨텍스트의 어떤 부분이 어려웠나요?

**Comment by byhhh2** [comment]
컴포넌트를 프로젝트안에서 특정 방식으로 나눈 이유 (Divider, FlexContainer, MarginContainer..)

**Comment by airman5573** [comment]
어려운 개념을 만났을때 어떻게 학습을 하시나요?

**Comment by airman5573** [comment]
자바스크립트에서 가장 흥미로운 부분은?

**Reply by airman5573** [reply]
실제적으로 쓰이는 기능중에서

**Comment by byhhh2** [comment]
callback 방식의 단점

**Reply by byhhh2** [reply]
제어권이란? (promise는 제어권을 어떻게 가져오나요?)

**Reply by byhhh2** [reply]
promise를 사용했을 때 예외처리를 어떻게 해주시나요?

**Reply by byhhh2** [reply]
promise의 단점?

**Comment by byhhh2** [comment]
type alias vs interface
